File: Extensions/BYTES.NET.Logging/Appenders/RollingFileAppender.vb
'import .net namespace(s) registered
Imports System.IO

'import namspace(s) required from 'BYTES.NET' library
Imports BYTES.NET.Logging.API

'import namespace(s) required from 'log4net' library
Imports log4net
Imports log4net.Repository.Hierarchy
Imports log4net.Layout
Imports log4net.Core

Namespace Appenders

    <LogAppenderMetadata(ID:="BYTES_LogAppender_RollingFile", Name:="Rolling File Log Appender")>
    Public Class RollingFileAppender

        Implements ILogAppender

#Region "private variable(s)"

        Private _filePath As String = Nothing

        Private _logger As ILog = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType)

        Private _loggingPattern As String = "%date %level [%thread] - %message%newline"
        Private _maxFileSize As String = "5MB"
        Private _maxBackupsCount As Integer = 3

#End Region

#Region "public properties"

        Public ReadOnly Property FilePath As String
            Get

                Return _filePath

            End Get
        End Property

#End Region

#Region "public new instance method(s)"

        ''' <summary>
        ''' default new instance method
        ''' </summary>
        ''' <param name="filePath"></param>
        ''' <param name="pattern"></param>
        ''' <param name="maxFileSize"></param>
        ''' <param name="maxBackupCounts"></param>
        Public Sub New(ByVal filePath As String, Optional ByVal pattern As String = Nothing, Optional ByVal maxFileSize As String = Nothing, Optional ByVal maxBackupCounts As Integer = Nothing)

            'create a new base-class instance
            MyBase.New

            'set the variable(s)
            _filePath = filePath

            If Not IsNothing(pattern) Then
                _loggingPattern = pattern
            End If

            If Not IsNothing(maxFileSize) Then
                _maxFileSize = maxFileSize
            End If

            If Not IsNothing(maxBackupCounts) Then
                _maxBackupsCount = maxBackupCounts
            End If

            'initialize 'log4net'
            ConfigureLog4Net()

        End Sub

#End Region

#Region "public method(s) inherited from base-class instance"

        ''' <summary>
        ''' method called on writing a log entry
        ''' </summary>
        ''' <param name="entry"></param>
        Public Sub Write(ByRef entry As LogEntry) Implements ILogAppender.Write

            Select Case (entry.Level)

                Case LogEntry.InformationLevel.Debug
                    _logger.Debug(entry.Message, entry.Details)

                Case LogEntry.InformationLevel.Info
                    _logger.Info(entry.Message, entry.Details)

                Case LogEntry.InformationLevel.Warning
                    _logger.Warn(entry.Message, entry.Details)

                Case LogEntry.InformationLevel.Exception
                    _logger.Error(entry.Message, entry.Details)

                Case LogEntry.InformationLevel.Fatal
                    _logger.Fatal(entry.Message, entry.Details)

            End Select

        End Sub

#End Region

#Region "private method(s)"

        ''' <summary>
        ''' method for configuring the 'log4net''RollingFileAppender' at runtime
        ''' </summary>
        ''' <remarks>based on the article found at 'https://stackoverflow.com/questions/16336917/can-you-configure-log4net-in-code-instead-of-using-a-config-file'</remarks>
        Private Sub ConfigureLog4Net()

            'check for the output folder (or create)
            Dim dirPath As String = _filePath.Substring(0, _filePath.LastIndexOf("\"))

            If Not Directory.Exists(dirPath) Then

                Directory.CreateDirectory(dirPath)

            End If

            'create the layout pattern
            Dim patternLayout As New PatternLayout()
            patternLayout.ConversionPattern = _loggingPattern
            patternLayout.ActivateOptions()

            'enable the rolling file appender
            Dim roller As Appender.RollingFileAppender = New Appender.RollingFileAppender()

            With roller
                .Layout = patternLayout

                .AppendToFile = True
                .File = _filePath
                .StaticLogFileName = True

                .RollingStyle = Appender.RollingFileAppender.RollingMode.Size
                .MaxSizeRollBackups = _maxBackupsCount
                .MaximumFileSize = _maxFileSize
            End With

            roller.ActivateOptions()

            'add the appender and configure the 'log4net' hirarchy
            Dim hierarchy As Hierarchy = DirectCast(LogManager.GetRepository(), Hierarchy)

            hierarchy.Root.AddAppender(roller)
            hierarchy.Root.Level = Level.All
            hierarchy.Configured = True

        End Sub

#End Region

    End Class

End Namespace

File: src-2.x/BYTES.NET.MS/IO/Logging/RollingFileAppender.cs
//import .net namespace(s) required
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BYTES.NET.IO.Logging;

//import namespace(s) required from 'BYTES.NET' framework
using BYTES.NET.IO.Logging.API;
using BYTES.NET.IO;

//import namespace(s) required from 'log4net' library
using log4net;
using log4net.Repository.Hierarchy;
using log4net.Layout;
using log4net.Core;

namespace BYTES.NET.MS.IO.Logging
{
    /// <summary>
    /// a rolling file appender
    /// </summary>
    /// <remarks>'Log4Net' wrapper class</remarks>
    public class RollingFileAppender : ILogAppender
    {

        #region private variable(s)

        private Log _parent = null;

        private string _filePath = null;
        private string _pattern = "%date %level [%thread] - %message%newline";
        private string _maxFileSize = "5MB";
        private int _maxBackupsCount = 3;
        private bool _dumpOnAttachement = true;

        private ILog _logger = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
        private bool _isAttached = false;

        #endregion

        #region public properties

        public string FilePath
        {
            get => _filePath;
            set
            {
                if (_isAttached)
                {
                    throw new InvalidOperationException("Unable to modify the file path after appender was attached");
                }

                _filePath = value;
            }
        }

        public string Pattern
        {
            get => _pattern;
            set
            {
                if (_isAttached)
                {
                    throw new InvalidOperationException("Unable to modify the pattern after appender was attached");
                }

                _pattern = value;
            }
        }

        public string MaxFileSize
        {
            get => _maxFileSize;
            set
            {
                if (_isAttached)
                {
                    throw new InvalidOperationException("Unable to modify the maximum file size after appender was attached");
                }

                _maxFileSize = value;
            }
        }

        public int MaxBackupsCount
        {
            get => _maxBackupsCount;
            set
            {
                if (_isAttached)
                {
                    throw new InvalidOperationException("Unable to modify the maximum number of backups after appender was attached");
                }

                _maxBackupsCount = value;
            }
        }

        public bool DumpOnAttachement
        {
            get => _dumpOnAttachement;
            set
            {
                if (_isAttached)
                {
                    throw new InvalidOperationException("Unable to modify the dumping behavior after appender was attached");
                }

                _dumpOnAttachement = value;
            }
        }

        #endregion

        #region public new instance method(s)

        /// <summary>
        /// default new instance method
        /// </summary>
        /// <param name="filePath"></param>
        /// <param name="args"></param>
        public RollingFileAppender(string filePath, Dictionary<string, object> args = null)
        {
            _filePath = filePath;

            if(args != null)
            {
                ParseArgs(args);
            }
        }

        #endregion

        #region public method(s) implementing 'ILogAppender'

        /// <summary>
        /// called on attachement
        /// </summary>
        /// <param name="log"></param>
        public void OnAppended(Log log)
        {
            _parent = log;

            //check for the output directory
            string dirPath = Path.GetDirectoryName(Helper.ExpandPath(_filePath));

            if (!Directory.Exists(dirPath))
            {
                Directory.CreateDirectory(dirPath);
            }

            //initialize the log
            InitializeLog4Net();
            _isAttached = true;

            //dump existing log entries
            if (_dumpOnAttachement)
            {
                _logger.Info("### start dumping log '" + _parent.ID + "' ###");

                foreach(LogEntry entry in _parent.GetCache())
                {
                    OnLogged(entry);
                }

                _logger.Info("### finished dumping log '" + _parent.ID + "' ###");
            }
        }

        /// <summary>
        /// called on logging
        /// </summary>
        /// <param name="entry"></param>
        /// <exception cref="NotImplementedException"></exception>
        public void OnLogged(LogEntry entry)
        {
            if (entry.Level == LogEntry.InformationLevel.Debug)
            {
                _logger.Debug(entry.Message, (Exception)entry.Details);
            }
            else if (entry.Level == LogEntry.InformationLevel.Info)
            {
                _logger.Info(entry.Message, (Exception)entry.Details);
            }
            else if (entry.Level == LogEntry.InformationLevel.Warning)
            {
                _logger.Warn(entry.Message, (Exception)entry.Details);
            }
            else if (entry.Level == LogEntry.InformationLevel.Exception)
            {
                _logger.Error(entry.Message, (Exception)entry.Details);
            }
            else if (entry.Level == LogEntry.InformationLevel.Fatal)
            {
                _logger.Fatal(entry.Message, (Exception)entry.Details);
            }
        }

        #endregion

        #region private method(s)

        /// <summary>
        /// parses the argument(s) given to internal variable(s)
        /// </summary>
        /// <param name="args"></param>
        private void ParseArgs(Dictionary<string,object> args)
        {
            foreach(KeyValuePair<string,object> arg in args)
            {
                if(arg.Key.ToLower() == "pattern")
                {
                    _pattern = (string)arg.Value;
                } else if(arg.Key.ToLower() == "maxfilesize")
                {
                    _maxFileSize = (string)arg.Value;
                } else if(arg.Key.ToLower() == "maxbackupscount")
                {
                    _maxBackupsCount = (int)arg.Value;
                } else if(arg.Key.ToLower() == "dumponattachement")
                {
                    _dumpOnAttachement=(bool)arg.Value;
                }
            }
        }

        /// <summary>
        /// initializes the 'Log4Net' framework
        /// </summary>
        /// <remarks>based on the article found at 'https://stackoverflow.com/questions/16336917/can-you-configure-log4net-in-code-instead-of-using-a-config-file'</remarks>
        private void InitializeLog4Net()
        {
            //validate the argument(s)
            if (_filePath == null || String.IsNullOrEmpty(_filePath) || String.IsNullOrWhiteSpace(_filePath))
            {
                throw new ArgumentException("The file path must not be empty");
            }

            //check for (and create) the output folder
            string filePath = Helper.ExpandPath(_filePath);
            string rootPath = filePath.Substring(0, filePath.LastIndexOf("\\"));

            if (!Directory.Exists(rootPath))
            {
                Directory.CreateDirectory(rootPath);
            }
            _filePath = filePath;

            //create the layout
            PatternLayout layout = new PatternLayout() { ConversionPattern = _pattern };
            layout.ActivateOptions();

            //enable the rolling file appender
            log4net.Appender.RollingFileAppender appender = new log4net.Appender.RollingFileAppender();
            appender.Layout = layout;
            appender.AppendToFile = true;
            appender.File = _filePath;
            appender.StaticLogFileName = true;
            appender.RollingStyle = log4net.Appender.RollingFileAppender.RollingMode.Size;
            appender.MaximumFileSize = _maxFileSize;
            appender.MaxSizeRollBackups = _maxBackupsCount;
            appender.ActivateOptions();

            //add the appender and configure the 'log4net' hierarchy
            Hierarchy hier = (Hierarchy)LogManager.GetRepository();
            hier.Root.AddAppender(appender);
            hier.Root.Level = Level.All;
            hier.Configured = true;
        }

        #endregion
    }
}
